### PR TITLE
feat(perf): add --ep-options to pass runtime EP provider options (#865)

### LINKED ISSUE #865: feat: add --ep-options to winml perf (and other commands) for runtime EP provider options
## Motivation

When benchmarking or running inference on QNN NPU (and other EPs), the runtime EP provider options can significantly affect latency — independently of the build-time quantization config. For example, on QNN HTP:

| EP option | Affects compile | Affects runtime | Values |
|---|---|---|---|
| \htp_performance_mode\ | ❌ (no-op) | ✅ clock governor | \urst\, \high_performance\, \alanced\, \low_power\, \default\ |
| \htp_graph_finalization_optimization_mode\ | ✅ changes compiled graph | ✅ | \

### PR DESCRIPTION
**Pull request by xieofxie**
## Summary

Fixes #865.

Adds a repeatable `--ep-options KEY=VALUE` flag to `winml perf` that forwards runtime execution-provider options to the inference session. These options (e.g. QNN HTP `htp_performance_mode`) significantly affect runtime latency independently of the build-time quantization config, so being able to set them at benchmark time is essential for tuning.

The flag works for **both** input modes:
- HuggingFace model IDs (`winml perf -m microsoft/resnet-50 --ep-options htp_performance_mode=burst`)
- Pre-exported ONNX files (`winml perf -m model.onnx --ep-options htp_performance_mode=burst`)

It is also wired into per-module (`--module`) benchmarking.

## How it works

The options are threaded down to `session.py`''s `add_ep_for_device`, which already accepts an `ep_options` dict:

```
--ep-options (CLI, repeatable)
  -> parse_ep_options() -> dict
  -> BenchmarkConfig.ep_options
  -> WinMLAutoModel.from_pretrained / from_onnx (provider_options=)
  -> WinMLPreTrainedModel (provider_options=)
  -> WinMLSession (provider_options=)
  -> _build_session_options -> add_ep_for_device(opts, ep, device_type, provider_options)
```

`WinMLSession` gains a `provider_options` kwarg that **merges on top of and overrides** any build-time `ep_config.provider_options`, but — unlike passing a full `ep_config` — does **not** flip EPContext persistence (`persist_jit`). These options tune the runtime session, not the compiled graph.

## Changes

- `utils/cli.py`: new `ep_options_option()` decorator and `parse_ep_options()` helper (reusable by other commands).
- `session/session.py`: `WinMLSession` accepts `provider_options`, merged over `ep_config` options.
- `models/winml/base.py`, `models/auto.py`: thread `provider_options` through `from_pretrained` / `from_onnx` (including the composite-model and skip-build paths).
- `commands/perf.py`: add `--ep-options`, parse once, pass to both single-model and `--module` paths.
- `docs/commands/perf.md`: document the new flag with an example.

## Tests

- `tests/unit/utils/test_cli.py`: `parse_ep_options` (empty/single/multiple/`=`-in-value/duplicate-key/invalid/empty-key).
- `tests/unit/commands/test_perf_cli.py`: `--ep-options` forwarded as `provider_options` for both ONNX and HF paths; CLI parsing into `BenchmarkConfig`; invalid-format rejection; help text.
- `tests/unit/session/test_winml_session.py`: runtime `provider_options` forwarded to `add_provider_for_devices`; runtime options override `ep_config` options.

All affected suites pass (225 passed; 2 pre-existing OpenVINO-EP-dependent failures unrelated to this change, deselected on this machine where OpenVINO is not installed).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/winml/modelkit/utils/cli.py [hunk 1]
```diff
@@ -187,6 +187,72 @@ def ep_option(required: bool = True, optional_message: str | None = None) -> Cal
...
+                param_hint="--ep-options",
+            )
+        key, value = item.split("=", 1)
+        key = key.strip()
+        if not key:
+            raise click.BadParameter(
+                f"Invalid EP option format: '{item}'. Key cannot be empty.",
+                param_hint="--ep-options",
+            )
+        options[key] = value
```
Comment (DingmaomaoBJTU) [inline src/winml/modelkit/utils/cli.py]: **Warning: value whitespace is not stripped, but key is.**

`key = key.strip()` guards the key, but `value` is assigned raw from `split('=', 1)`. Passing `--ep-options 'htp_performance_mode= burst'` or `'htp_performance_mode=burst '` (trailing space from shell quoting) will silently forward a wrong value to ORT - likely causing the option to have no effect and being very hard to debug.

Suggestion: `options[key] = value.strip()`

File: src/winml/modelkit/utils/cli.py [hunk 2]
```diff
@@ -187,6 +187,72 @@ def ep_option(required: bool = True, optional_message: str | None = None) -> Cal
...
+        Mapping of option name to value, or ``None`` when nothing was provided
+        (so callers can leave the session default untouched).
+
+    Raises:
+        click.BadParameter: If any value is missing the ``=`` separator or has
+            an empty key.
+    """
+    if not values:
+        return None
+    options: dict[str, str] = {}
```
Comment (DingmaomaoBJTU) [inline src/winml/modelkit/utils/cli.py:242]: **Info: duplicate-key last-write-wins is tested but undocumented.**

`--ep-options k=1 --ep-options k=2` silently uses `k=2`. Reasonable semantics, but neither the help text nor the docstring mentions it. A one-liner in the docstring (e.g. 'Duplicate keys: later occurrence wins.') would prevent confusion.

File: src/winml/modelkit/commands/perf.py [hunk 3]
```diff
@@ -84,6 +84,7 @@ class BenchmarkConfig:
     allow_unsupported_nodes: bool = False
     monitor: bool = False
     ep: EPNameOrAlias | None = None
+    ep_options: dict[str, str] | None = None
```
Comment (DingmaomaoBJTU) [inline src/winml/modelkit/commands/perf.py:87]: **Info: `ep_options` is not captured in `BenchmarkResult.to_dict()`.**

`BenchmarkConfig.ep_options` is correctly forwarded to ORT but never written into the `benchmark_info` block of the JSON report. A saved `.json` result is therefore not fully reproducible - you cannot re-run the exact benchmark from it alone.

Consider adding `\\p_options\\: self.config.ep_options` to the `benchmark_info` dict in `to_dict()`.

**Review by DingmaomaoBJTU — COMMENTED** [review]
Overall this PR is well-structured: parsing, session forwarding, and test coverage are solid. Three items worth discussing below.

File: src/winml/modelkit/commands/perf.py (hunk 3 above)
Comment (xieofxie) [reply src/winml/modelkit/commands/perf.py:87]: added

File: src/winml/modelkit/utils/cli.py (hunk 1 above)
Comment (xieofxie) [reply src/winml/modelkit/utils/cli.py]: added

File: src/winml/modelkit/utils/cli.py (hunk 2 above)
Comment (xieofxie) [reply src/winml/modelkit/utils/cli.py:242]: added

File: src/winml/modelkit/utils/cli.py [hunk 4]
```diff
@@ -187,6 +187,76 @@ def ep_option(required: bool = True, optional_message: str | None = None) -> Cal
...
+                param_hint="--ep-options",
+            )
+        key, value = item.split("=", 1)
+        key = key.strip()
+        if not key:
+            raise click.BadParameter(
+                f"Invalid EP option format: '{item}'. Key cannot be empty.",
+                param_hint="--ep-options",
+            )
+        options[key] = value.strip()
```
Comment (DingmaomaoBJTU) [inline src/winml/modelkit/utils/cli.py:256]: **Resolved (was WARNING):** `options[key] = value.strip()` now strips whitespace from both key and value. The new test `test_whitespace_stripped_from_key_and_value` confirms edge cases including keys with surrounding spaces. :white_check_mark:

File: src/winml/modelkit/utils/cli.py [hunk 5]
```diff
@@ -187,6 +187,76 @@ def ep_option(required: bool = True, optional_message: str | None = None) -> Cal
...
+    Args:
+        optional_message: Extra command-specific guidance appended to help text.
+
+    Returns:
+        Decorator function.
+    """
+    help_text = (
+        "Runtime EP provider option as KEY=VALUE (repeatable). Forwarded to the "
+        "inference session's execution provider (e.g. "
+        "--ep-options htp_performance_mode=burst). Duplicate keys: later "
```
Comment (DingmaomaoBJTU) [inline src/winml/modelkit/utils/cli.py:209]: **Resolved (was INFO):** Duplicate-key last-write-wins semantics are now documented in both the `--help` text (line 209-210) and the `parse_ep_options` docstring (line 229). :white_check_mark:

File: src/winml/modelkit/commands/perf.py [hunk 6]
```diff
@@ -147,6 +148,7 @@ def to_dict(self) -> dict[str, Any]:
                 "task": self.actual_task,
                 "device": self.actual_device,
                 "ep": self.actual_ep,
+                "ep_options": self.config.ep_options,
```
Comment (DingmaomaoBJTU) [inline src/winml/modelkit/commands/perf.py:151]: **Resolved (was INFO):** `ep_options` is now written into `benchmark_info` in `to_dict()`, and two new tests (`test_ep_options_captured_in_to_dict`, `test_ep_options_none_when_not_set_in_to_dict`) verify the behaviour. :white_check_mark:

File: src/winml/modelkit/session/session.py [hunk 7]
```diff
@@ -391,11 +401,12 @@ def _build_session_options(self, device: str) -> tuple[ort.SessionOptions, str,
         opts = self._session_options_factory()
         if add_ep_for_device(opts, resolved_ep, device_type, self._provider_options):
             logger.info(
-                "Built SessionOptions with EP: %s (%s) device=%s -> %s",
+                "Built SessionOptions with EP: %s (%s) device=%s -> %s provider_options=%s",
```
Comment (DingmaomaoBJTU) [inline src/winml/modelkit/session/session.py:404]: **New (INFO):** The updated log line emits `self._provider_options` at `INFO` level. EP provider options are typically performance knobs (e.g. `htp_performance_mode=burst`) rather than secrets, so this is low risk. Worth being aware of in case a future EP expects a token or API key as a provider option — in that case `DEBUG` level would be more appropriate. No action required now.

**Review by DingmaomaoBJTU — COMMENTED** [review]
Follow-up review: all three previous findings have been addressed. One minor new observation below.

**Review by DingmaomaoBJTU — APPROVED** [review]
All previous findings addressed. LGTM :+1:

File: src/winml/modelkit/session/session.py (hunk 7 above)
Comment (xieofxie) [reply src/winml/modelkit/session/session.py:404]: so careful
